Prototype of new Easy, Normal, Hard, and Brutal AIs for RA Mod.

blackhand1001 opened the pull request:
Prototype of new Easy, Normal, Hard, and Brutal AIs. This includes several fixes the the bot modules to reduce the glitchy behavior.
-Added new targeting code to select the nearest building for attack move rather than nearest unit which is often an aircraft or harvester. If no buildings exist it will revert to selecting the nearest unit.
-Remove the regrouping code from naval squads as many maps have the water around the border which means that the leader unit may technically closer to the base its actual has a further distance to travel so the squad endlessly drives back towards it.
-Fix the repairbot module to repair buildings if they have any damage rather than waiting for them to be below 75%.
-Hard and brutal AI bots will build a second MCV to expand their base area so that they don't run out of room.
-Change default AI stance on subs to be default (defend) instead of return fire. Player stance is unchanged.
-Add in slight money trickles for hard and brutal AI. This is tied to the refineries and uses the resource storage.

This resolves issue #16126 for the RA mod. 

Some features of this rely on PR's #17309 #17236 #17276 #17298. It will function without those but the intended behavior includes those PRs.

blackhand1001 committed:
Prototype of new Easy, Normal, Hard, and Brutal AIs for RA Mod

Prototype of new Easy, Normal, Hard, and Brutal AIs for RA Mod. This includes several fixes the the bot modules to reduce the glitchy behavior.
-Added new targetting code to select the nearest building for attack move rather than nearest unit which is often an aircraft or harvester. If no buildings exist it will revert to selecting the nearest unit.
-Remove the regrouping code from naval squads as many maps have the water around the border which means that the leader unit may technically closer to the base its actual has a further distance to travel so the squad endlessly drives back towards it.
-Fix the repairbot module to repair buildings if they have any damage rather than waiting for them to be below 75%.
-Hard and brutal AI bots will build a second MCV to expand their base area so that they don't run out of room.
-Change default AI stance on subs to be default (defend) instead of return fire. Player stance is unchanged.
-Add in slight money trickles for hard and brutal AI. This is tied to the refineries and uses the resource storage.

GraionDilach commented:
> -Fix the repairbot module to repair buildings if they have any damage rather than waiting for them to be below 75%.

Doesn't this break Dune 2000 AIs?

blackhand1001 commented:
No. Repairs don't occur until after an enemy unit damages the building.

matjaeck commented:
> We are planning to rework the AI in the near future anyway, so i'm not sure if adding this now to be later removed/replaced is a good idea. See #16126 and  #16634.

_Originally posted by @MustaphaTR in https://github.com/OpenRA/OpenRA/pull/17213#issuecomment-540994888_

I also wonder if it makes sense to merge the rules changes, especially the ones related to production when we have #16633 in the pipeline. The AI refactor is not within the scope of  #17307 so we don't need to rush it now, and merging rules changes that - provided everything works as planned - won't make into a stable release is not really helpful, condidering that it needs to be reviewed though. The new production considerations and potentially a build order trait I'm working on will change things a lot, so the cash trickler for example which I'm not a big fan of might not be required anymore then.

Since activity has slowed down it is not unlikely that the PR will take quite a while to get merged, potentially until the other refactors are done, and it would be a shame if those completely superseded this PR then.

So IMO it makes sense to a) separate commits in code(fixes) and rules changes and b) limit the scope of the rules changes to bugfixes or general improvements that won't get overwritten in the forseeable future. Best would be IMO to limit any short term AI changes to bug fixes like those in #17307 and defer everything else to when we have all the things ready that are needed for the refactor (mainly #16633 and required traits for game time and income).

matjaeck commented:
I updated my previous comment multiple times, so thought I should note that in a new comment to avoid misunderstandings.

blackhand1001 commented:
Here's a demonstration of the improved ai in action. This is a hard Russia vs Brutal England. Tickrate is a little above stock speed. Notice that naval works properly now, and that at the end the naval units finish off the soviet base. 
https://www.youtube.com/watch?v=iDBEuUJbcI0

blackhand1001 commented:
I need help. I rebased this on the latest bleed and now its showing a bunch of extra commits and file changes.

chrisforbes commented on OpenRA.Mods.Common/Traits/BotModules/BuildingRepairBotModule.cs:
```diff
@@ -35,7 +35,7 @@ void IBotRespondToAttack.RespondToAttack(IBot bot, Actor self, AttackInfo e)
 			if (rb != null)
 			{
-				if (e.DamageState > DamageState.Light && e.PreviousDamageState <= DamageState.Light && !rb.RepairActive)
+				if (e.DamageState > DamageState.Undamaged && e.PreviousDamageState <= DamageState.Undamaged && !rb.RepairActive)
```
This is going to make the AI obsess over keeping its building health topped up... is that right?

blackhand1001 committed:
If enemy builds their first naval yard since last tick, invalidate navysquads current target and select one near it as its more reliable.

If enemy builds their first naval yard since last tick, invalidate navysquads current target and select one near it as its more reliable.

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/NavyStates.cs:
```diff
@@ -100,7 +101,17 @@ public void Tick(Squad owner)
+			var navalProductions = owner.World.ActorsHavingTrait<Building>().Where(a
+			=> owner.SquadManager.Info.NavalProductionTypes.Contains(a.Info.Name)
+			&& domainIndex.IsPassable(first.Location, a.Location, locomotorInfo)
+			&& a.AppearsHostileTo(first));
```
lines 109-111 should receive one mor level of indentation to make it more visible that they're par of `var navalProductions`.

blackhand1001 replied:
Will do.

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/NavyStates.cs:
```diff
@@ -100,7 +101,17 @@ public void Tick(Squad owner)
 				return;
 
-			if (!owner.IsTargetValid)
+			// Check if we have an enemy naval yard this tick
```
I'm a bit uncomfortable about checking all this every tick, for performance reasons.
I think it would be better if the squad only checked every second (25 ticks) or so.

blackhand1001 replied:
From my understanding of the squad code, this tick only called on the AttackForceInterval. Its called by the fuzzystatemachine which only updates when AttackForceTick <= 0 in squadmanager. So by default its only checking every 75 ticks.

matjaeck commented:
Deleted my objections since they were ignoted anyway - let's take this for the AI overhaul theme then.

blackhand1001 committed:
Add UnitBuilderInterval trait

Add UnitBuilderInterval trait to set the interval between checking to build units with available queues.

blackhand1001 committed:
Add HarvestersPerRefinery trait to HarvesterBotModule

Controls the number of harvesters to build for each refinery for the harvester replacement system.

GraionDilach commented:
Not too fond of the cashtricklers, it's a very lousy way of money cheat. Go with what RA2 did and add ResourcePurifier traits to the AI definitions, enabled with AI conditions instead.

blackhand1001 commented:
Rebased this and squashed a bunch of commits together but kept the new traits each as their own seperate commit.

blackhand1001 committed:
Reduce protectionscan radius a bit.

blackhand1001 commented:
> Not too fond of the cashtricklers, it's a very lousy way of money cheat. Go with what RA2 did and add ResourcePurifier traits to the AI definitions, enabled with AI conditions instead.

I actually did the cashtrickers on the refs themselves so that destroying refs slows down their income and if you kill all of the refs and silos they will have no money left.

blackhand1001 committed:
Dont Add naval or aircraft to protectsquads

Aircraft have their own base patrol code, and naval units can't pathfind to most places in the base and therefore end up stuck idle.

blackhand1001 committed:
Tweak protectunitscanradius a bit

Tweak protectunitscanradius a bit

blackhand1001 committed:
Update GroundStates.cs

blackhand1001 committed:
Update StateBase.cs

blackhand1001 committed:
Revert "Update StateBase.cs"

This reverts commit 4d2fc311c6c43fd22a677fce390cddf6987a0b9c.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update ai.yaml

GraionDilach commented:
> > Not too fond of the cashtricklers, it's a very lousy way of money cheat. Go with what RA2 did and add ResourcePurifier traits to the AI definitions, enabled with AI conditions instead.
> 
> I actually did the cashtrickers on the refs themselves so that destroying refs slows down their income and if you kill all of the refs and silos they will have no money left.

It still looks way more fake than having virtual purifiers though.

If you use virtual purifiers, the economy graph keeps the shape it would have without a moneycheat, with only having a scale difference. With cashtrickler-based methods, the economy graph ends up having an offset, also leading to the harvester relegated into a secondary income.

This also means harvester sniping becomes useless against AIs while virtual purifiers would mean the AI can be forced to run out of money with having it's harvesters denied constantly.

blackhand1001 committed:
Update GroundStates.cs

blackhand1001 committed:
When forming attack forces and rush attacks, remove units from protectsquad

When forming attack forces and rush attacks, remove units from protectsquad

blackhand1001 committed:
Update ai.yaml

abcdefg30 commented:
What is the reasoning behind cheating money for the AI? I didn't have the impression our old AI did that bad money wise. Changing established game rules arbitrarily sounds like a bad idea to me.

pchote commented:
Adding my vote to the "multiply resource value, don't trickle unrelated funds" bandwagon. This was in fact one of the specific points outlined in #16126.

GraionDilach commented:
>What is the reasoning behind cheating money for the AI? I didn't have the impression our old AI did that bad money wise.

The AI has a really bad habit at picking up lost harvesters and/or starting to build refineries really late, being too prone to eco sniping. Or it just ends up overspending (moreso in TD than in RA just due to parallel queues). Moneycheats are proposed to mitigate this and considering that moneycheats were in all Westwood games (TD/RA/TS added a starting money bonus, RA2 debuted with virtual purifiers).

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Queue the goto Random Building orders

Queue the goto Random Building orders instead of doing them instantly in case we are finishing an action.

abcdefg30 commented:
I think arbitrary changes only for the AI are bad since they go against what players know about and expect from certain gameplay situations. Just doing this because the originals did it doesn't seem to be a good argument either. However, only doing it for the hardest AI (as proposed in #16126) is okay I suppose.

blackhand1001 commented:
Only hard and brutal have the money cheat. Easy and normal do not cheat.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Add UnitBuilderMinCredits trait.

Add UnitBuilderMinCredits trait that disables the pauses the unitbuilder when credits go below a set amount.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update ai.yaml

GraionDilach commented:
Again, please remove the cashtricklers and use purifiers instead.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update NavyStates.cs

matjaeck commented:
Please let us know when you're happy / done with this so testing can start.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Remove old obsolete turtle and rush AIs.

Remove old obsolete turtle and rush AIs.

blackhand1001 commented:
This should be at the point where its testable. Feel free to build it and play against it. The more feedback I can get, the better.

blackhand1001 committed:
Revert "Remove old obsolete turtle and rush AIs."

This reverts commit 5673aefd6bae235599f4a14626b2fdefaec9f6be.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update ai.yaml

matjaeck commented:
Note that it was just a general comment. I won't have time for any testing in the future, so others will need to do this.

reaperrr commented:
Even if testing may be possible already, the PR has become quite large and hard to read.

For starters, could you split out commits 4c2bc0d and 3f15980 (protection squad fixes) to a separate PR, and 02fdbef (GoToRandomOwnBuilding fix) to another, 2nd separate PR?

Those fixes are straightforward and small enough they could be tested and merged fairly quickly, and are good candidates for a bugfix release.

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update GroundStates.cs

blackhand1001 committed:
Update ai.yaml

blackhand1001 committed:
Update AirStates.cs

blackhand1001 committed:
Update ai.yaml

pchote commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:
```diff
@@ -23,7 +23,7 @@ protected static void GoToRandomOwnBuilding(Squad squad)
 			var loc = RandomBuildingLocation(squad);
 			foreach (var a in squad.Units)
-				squad.Bot.QueueOrder(new Order("Move", a, Target.FromCell(squad.World, loc), false));
+				squad.Bot.QueueOrder(new Order("Move", a, Target.FromCell(squad.World, loc), true));
```
Quoting my review from #17408 so its not forgotten here:

> `GoToRandomOwnBuilding` is called when units transition from an attack to a flee state. Canceling the current activity so the units can run away is a core part of the originally designed logic.
>
> Are we happy to be intentionally breaking this?
>
> IMO it is reasonable to reconsider whether the attack/flee system makes sense at all, but if we decide that it doesn't we need to remove/rewrite it properly.
>
> If the goal is to reduce the probability / raise the danger threshold for units running away we should be looking at the logic inside the `ShouldFlee` methods instead!

NotoriousPyro commented:
Please do not make the AI cheat. That is one of the most annoying 'features' of the likes of Tiberian Sun as an attempt to make up for poorer AI.

Other avenues of making the AI make smarter choices, such as building new con yards nearer fresh tiberium fields comes to mind ABOVE allowing cheats. I've just cloned patch-1 and applied it onto the current OpenRA branch, it is most certainly not a 'slow trickle' - it's completely unfair to the player and should be renamed to 'Cheat AI' if it's to be included at all...

![image](https://user-images.githubusercontent.com/706132/73692114-9af0d000-46cb-11ea-9cf2-a0e5e8387395.png)

NotoriousPyro commented:
Example, here is an enhancement to make the AI MCV placement choice to be based on resources which are outside of an MCV owned by the current player. So the AI actively looks for more resources instead of cramming the base into one place.

It could probably be further enhanced to avoid enemy bases too...

```
public CPos GetRandomBaseCenter(bool distanceToBaseIsImportant)
{
	if (distanceToBaseIsImportant == false)
		return initialBaseCenter;

	var tileset = world.Map.Rules.TileSet;
	BitArray resourceTypeIndices;

	resourceTypeIndices = new BitArray(tileset.TerrainInfo.Length);
	foreach (var t in world.Map.Rules.Actors["world"].TraitInfos<ResourceTypeInfo>())
		resourceTypeIndices.Set(tileset.GetTerrainIndex(t.TerrainType), true);

	var randomConstructionYard = world.Actors.Where(a => a.Owner == player &&
		Info.ConstructionYardTypes.Contains(a.Info.Name))
		.RandomOrDefault(world.LocalRandom);

	var newResources = world.Map.FindTilesInAnnulus(randomConstructionYard.Location, Info.MaxBaseRadius, world.Map.Grid.MaximumTileSearchRange)
		.Where(a => resourceTypeIndices.Get(world.Map.GetTerrainIndex(a)))
		.Shuffle(world.LocalRandom).FirstOrDefault();

	return newResources;
}
```

Example in use with the cheating AI from this PR makes the AI more aggressive with MCV placement and expansion:
![image](https://user-images.githubusercontent.com/706132/73702798-d1d3df80-46e5-11ea-90b6-45a10d9ebe83.png)

KOYK commented:
ehem so @NotoriousPyro are you going to make a PR with your solution or if you don't have the time could some one else do it? since your idea obviously is the way to go, and also move this project's AI a step forward from stone age.

pchote commented:
This is not a good time to be trying to do anything with the AI. We need time to focus on improving some of the fundamental issues with the AI code, and new features distract and make this work more difficult.

KOYK commented:
So I guess its up to @NotoriousPyro to fully implement this alone.

pchote commented:
There doesn't seem to be much more we can do here, so moving to the Future milestone so we can continue to pick ideas from this, and closing to help clean up the PR queue.
